include/aws/s3/private/s3_client_impl.h

```diff
@@ -26,24 +26,39 @@ typedef void(aws_s3_client_get_http_connection_callback)(
 
 typedef void(aws_s3_client_sign_callback)(int error_code, void *user_data);
 
+typedef void(aws_s3_vip_shutdown_callback_fn)(void *user_data);
+
 /* Represents one Virtual IP (VIP) in S3, including a connection manager that points directly at that VIP. */
 struct aws_s3_vip {
     struct aws_linked_list_node node;
 
-    /* True if this VIP is currently being cleaned up.  The work event loop will check for this flag and clean up
-     * related VIP connections. */
+    /* True if this VIP is in use. */
     struct aws_atomic_var active;
 
-    struct aws_ref_count internal_ref_count;
-
     /* S3 Client that owns this vip. */
     struct aws_s3_client *owning_client;
 
+    /* Connection manager shared by all VIP connections. */
+    struct aws_http_connection_manager *http_connection_manager;
+
     /* Address this VIP represents. */
     struct aws_string *host_address;
 
-    /* Connection manager shared by all VIP connections. */
-    struct aws_http_connection_manager *http_connection_manager;
+    /* Callback used when this vip has completely shutdown, which happens when all associated connections and the
+     * connection manager are shutdown. */
+    aws_s3_vip_shutdown_callback_fn *shutdown_callback;
+
+    /* User data for the shutdown callback. */
+    void *shutdown_user_data;
+
+    struct {
+        /* How many vip connections are allocated for this vip. */
+        uint32_t num_vip_connections;
```

Comment:
Is this always in sync with the connection manager? Should it be? at a glance, I can't tell why it's separate.

Reply:
Ok I figured it out. Maybe just note that this is how many are active. A difference between this and the connection manager/0 means we can't shutdown yet, or something like that?